src/gt4py/next/iterator/transforms/fuse_maps.py

```diff
@@ -0,0 +1,131 @@
+    """
+
+    uids: UIDGenerator = dataclasses.field(init=False, repr=False, default_factory=UIDGenerator)
+
+    def _as_lambda(self, fun: ir.SymRef | ir.Lambda, param_count: int) -> ir.Lambda:
+        # if fun is already a Lambda we still wrap it to get unique symbol names to avoid symbol clashes
+        params = [
+            ir.Sym(id=self.uids.sequential_id(prefix="_fuse_maps")) for _ in range(param_count)
+        ]
+        return ir.Lambda(
+            params=params,
+            expr=ir.FunCall(fun=fun, args=[ir.SymRef(id=p.id) for p in params]),
+        )
+
+    # TODO think about clashes of symbol names
+    def visit_FunCall(self, node: ir.FunCall, **kwargs):
+        node = self.generic_visit(node)
+        if _is_map(node) or _is_reduce(node):
+            if any(_is_map(arg) for arg in node.args):
+                first_param = (
+                    0 if _is_map(node) else 1
+                )  # index of the first param of op that maps to args (0 for map, 1 for reduce)
+                assert isinstance(node.fun, ir.FunCall)
+                assert isinstance(node.fun.args[0], (ir.Lambda, ir.SymRef))
+                outer_op = self._as_lambda(node.fun.args[0], len(node.args) + first_param)
+
+                inlined_args = []
+                new_params = []
+                new_args = []
+                if _is_reduce(node):
+                    # param corresponding to reduce acc
+                    inlined_args.append(ir.SymRef(id=outer_op.params[0].id))
+                    new_params.append(outer_op.params[0])
+
+                for i in range(len(node.args)):
+                    if _is_map(node.args[i]):
+                        map_call = node.args[i]
+                        assert isinstance(map_call, ir.FunCall)
+                        assert isinstance(map_call.fun, ir.FunCall)
+                        assert isinstance(map_call.fun.args[0], (ir.Lambda, ir.SymRef))
+                        inner_op = self._as_lambda(map_call.fun.args[0], len(map_call.args))
+                        inlined_args.append(
+                            inline_lambdas.inline_lambda(
+                                ir.FunCall(
+                                    fun=inner_op,
+                                    args=[ir.SymRef(id=param.id) for param in inner_op.params],
+                                )
+                            )
+                        )
+                        new_params.extend(inner_op.params)
+                        new_args.extend(map_call.args)
+                    else:
+                        inlined_args.append(ir.SymRef(id=outer_op.params[i + first_param].id))
+                        new_params.append(outer_op.params[i + first_param])
+                        new_args.append(node.args[i])
+                new_body = ir.FunCall(
+                    fun=outer_op,
+                    args=inlined_args,
+                )
+                new_body = inline_lambdas.InlineLambdas.apply(new_body)
```

Comment:
TODO comment why we need the recursive version

Reply:
TODO test if we really need the inlining for unroll_reduce (if not we can rely on the final inline_lambda in pass_manager).

Reply:
the inlining is not needed for unroll_reduce, therefore I reverted back to the non-recursive inliner here. Instead I added the recursive inliner in the tests to have simpler and more stable patterns.